crypto/s2n_rsa_pss.c

```diff
@@ -0,0 +1,305 @@
+    S2N_ERROR_IF(signature_len > signature_out->size, S2N_ERR_SIZE_MISMATCH);
+
+    /* Actually sign the the digest */
+    GUARD_OSSL(EVP_PKEY_sign(ctx, signature_out->data, &signature_len, digest_data, digest_length), S2N_ERR_SIGN);
+    signature_out->size = signature_len;
+
+    return 0;
+}
+
+int s2n_rsa_pss_verify(const struct s2n_pkey *pub, struct s2n_hash_state *digest, struct s2n_blob *signature_in)
+{
+    notnull_check(pub);
+
+    /* Using Private Key to Verify means the public/private keys were likely swapped, and likely indicates a bug. */
+    S2N_ERROR_IF(s2n_rsa_is_private_key(pub->key.rsa_pss_key.pkey), S2N_ERR_KEY_MISMATCH);
+
+    uint8_t digest_length;
+    uint8_t digest_data[S2N_MAX_DIGEST_LEN];
+    GUARD(s2n_hash_digest_size(digest->alg, &digest_length));
+    GUARD(s2n_hash_digest(digest, digest_data, digest_length));
+    const EVP_MD* digest_alg = s2n_hash_alg_to_evp_alg(digest->alg);
+
+    /* For more info see: https://www.openssl.org/docs/manmaster/man3/EVP_PKEY_verify.html */
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(pub->key.rsa_pss_key.pkey, NULL);
+    notnull_check(ctx);
+
+    GUARD_OSSL(EVP_PKEY_verify_init(ctx), S2N_ERR_VERIFY_SIGNATURE);
+    GUARD_OSSL(EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PSS_PADDING), S2N_ERR_SIGN);
+    GUARD(s2n_evp_ctx_set_signature_digest(ctx, digest_alg, S2N_ERR_VERIFY_SIGNATURE));
+    GUARD_OSSL(EVP_PKEY_verify(ctx, signature_in->data, signature_in->size, digest_data, digest_length), S2N_ERR_VERIFY_SIGNATURE);
+
+    return 0;
+}
+
+static int s2n_rsa_pss_validate_sign_verify_match(const struct s2n_pkey *pub, const struct s2n_pkey *priv) {
+    /* Generate a random blob to sign and verify */
+    s2n_stack_blob(random_data, RSA_PSS_SIGN_VERIFY_RANDOM_BLOB_SIZE, RSA_PSS_SIGN_VERIFY_RANDOM_BLOB_SIZE);
+    GUARD(s2n_get_private_random_data(&random_data));
+
+    /* Sign/Verify API's only accept Hashes, so hash our Random Data */
+    DEFER_CLEANUP(struct s2n_hash_state sign_hash = {0}, s2n_hash_free);
+    DEFER_CLEANUP(struct s2n_hash_state verify_hash = {0}, s2n_hash_free);
+    GUARD(s2n_hash_new(&sign_hash));
+    GUARD(s2n_hash_new(&verify_hash));
+    GUARD(s2n_hash_init(&sign_hash, S2N_HASH_SHA256));
+    GUARD(s2n_hash_init(&verify_hash, S2N_HASH_SHA256));
+    GUARD(s2n_hash_update(&sign_hash, random_data.data, random_data.size));
+    GUARD(s2n_hash_update(&verify_hash, random_data.data, random_data.size));
+
+    /* Sign and Verify the Hash of the Random Blob */
+    s2n_stack_blob(signature_data, RSA_PSS_SIGN_VERIFY_SIGNATURE_SIZE, RSA_PSS_SIGN_VERIFY_SIGNATURE_SIZE);
+    GUARD(s2n_rsa_pss_sign(priv, &sign_hash, &signature_data));
+    GUARD(s2n_rsa_pss_verify(pub, &verify_hash, &signature_data));
+
+    return 0;
+}
+
+static int s2n_rsa_validate_param_equal(const RSA *pub, const RSA *priv, int required,
+                                        ossl_get_rsa_pss_param_fn get_rsa_pss_param_fn) {
```

Comment:
Parameter name is still wrong.

Maybe `osss_get_rsa_bignum_fn`?

Reply:
Fixed